### ISSUE TITLE
Can't compile - undefined reference to `Laxkit::RefPtrStack<LaxInterfaces::anInterface>::RefPtrStack(char)

### ISSUE THREAD

**Issue by madsl** [Laidout/laidout#23]
Tried for a while now to create an ebuild for this for Gentoo, but I cannot get this to compile on any compiler I have available, it always ends up with this error:

```
g++ core/document.o core/objectiterator.o core/drawdata.o core/fieldplace.o core/guides.o core/importimage.o core/laidoutprefs.o core/page.o core/papersizes.o core/plaintext.o core/project.o core/spreadview.o core/stylemanager.o core/utils.o calculator/values.o calculator/calculator.o calculator/interpreter.o calculator/curvevalue.o calculator/shortcuttodef.o impositions/imposition.o impositions/impositioneditor.o impositions/singles.o impositions/impositions.o impositions/dodecahedron.o impositions/box.o impositions/netimposition.o impositions/simplenet.o impositions/signatures.o impositions/signatureinterface.o impositions/netdialog.o impositions/polyptychwindow.o impositions/singleseditor.o plugins/plugin.o api/reimpose.o api/openandnew.o api/importexport.o api/buildicons.o api/functions.o interfaces/actionarea.o interfaces/nupinterface.o interfaces/aligninterface.o interfaces/anchorinterface.o interfaces/pagerangeinterface.o interfaces/graphicalshell.o interfaces/objectindicator.o interfaces/paperinterface.o interfaces/cloneinterface.o interfaces/animationinterface.o interfaces/pagemarkerinterface.o interfaces/objectfilterinterface.o interfaces/pathintersectionsinterface.o nodes/nodeinterface.o nodes/nodeeditor.o nodes/nodes-dataobjects.o nodes/nodes.o printing/print.o printing/psfilters.o printing/pscolorpatch.o printing/psgradient.o printing/psimage.o printing/psimagepatch.o printing/pspathsdata.o printing/pseps.o printing/epsutils.o printing/psout.o filetypes/filefilters.o filetypes/filters.o filetypes/exportdialog.o filetypes/htmlgallery.o filetypes/importdialog.o filetypes/image.o filetypes/image-gs.o filetypes/laidoutimport.o filetypes/pageatlas.o filetypes/pdf.o filetypes/podofoimposeplan.o filetypes/postscript.o filetypes/ppt.o filetypes/scribus.o filetypes/svg.o dataobjects/affinevalue.o dataobjects/pointsetvalue.o dataobjects/bboxvalue.o dataobjects/group.o dataobjects/objectcontainer.o dataobjects/objectfilter.o dataobjects/drawableobject.o dataobjects/datafactory.o dataobjects/groupinterface.o dataobjects/epsdata.o dataobjects/fontvalue.o dataobjects/helpertypes.o dataobjects/imagevalue.o dataobjects/limagepatch.o dataobjects/limagedata.o dataobjects/limageinterface.o dataobjects/lgradientdata.o dataobjects/lperspectiveinterface.o dataobjects/lpathsdata.o dataobjects/mysterydata.o dataobjects/lsomedataref.o dataobjects/lengraverfilldata.o dataobjects/lcaptiondata.o dataobjects/ltextonpath.o dataobjects/lvoronoidata.o dataobjects/printermarks.o ui/about.o ui/autosavewindow.o ui/buttonbox.o ui/commandwindow.o ui/findwindow.o ui/headwindow.o ui/helpwindow.o ui/metawindow.o ui/importimagesdialog.o ui/newdoc.o ui/objecttree.o ui/palettes.o ui/papersizewindow.o ui/plaintextwindow.o ui/pluginwindow.o ui/settingswindow.o ui/spreadeditor.o ui/viewwindow.o ui/valuewindow.o ui/interfaces.o  laidout-more.o laidout.o   polyptych/src/nets.o polyptych/src/poly.o -L/var/tmp/portage/app-editors/laidout-9999/work/laidout-9999/laxkit/lax/interfaces -L/var/tmp/portage/app-editors/laidout-9999/work/laidout-9999/laxkit/lax -llaxinterfaces -llaxkit -lharfbuzz -lcairo -lfontconfig  -L/usr/local/lib -L/usr/X11R6/lib -rdynamic -lXi -lXext -lX11 -lm -lpng `imlib2-config --libs` `pkg-config --libs freetype2` `pkg-config --libs GraphicsMagick++` `cups-config --libs` -ldl -lXft -lXrandr  -lreadline -lcrypto  -o laidout
/usr/lib/gcc/x86_64-pc-linux-gnu/6.5.0/../../../../x86_64-pc-linux-gnu/bin/ld: ui/interfaces.o: in function `Laidout::GetBuiltinInterfaces(Laxkit::RefPtrStack<LaxInterfaces::anInterface>*)':
/var/tmp/portage/app-editors/laidout-9999/work/laidout-9999/src/ui/interfaces.cc:95: undefined reference to `Laxkit::RefPtrStack<LaxInterfaces::anInterface>::RefPtrStack(char)'
/usr/lib/gcc/x86_64-pc-linux-gnu/6.5.0/../../../../x86_64-pc-linux-gnu/bin/ld: /var/tmp/portage/app-editors/laidout-9999/work/laidout-9999/laxkit/lax/interfaces/liblaxinterfaces.a(textonpathinterface.o): in function `LaxInterfaces::TextOnPath::ConvertToPaths(bool, Laxkit::RefPtrStack<LaxInterfaces::SomeData>*)':
/var/tmp/portage/app-editors/laidout-9999/work/laidout-9999/laxkit/lax/interfaces/textonpathinterface.cc:986: undefined reference to `Laxkit::RefPtrStack<LaxInterfaces::SomeData>::RefPtrStack(char)'
/usr/lib/gcc/x86_64-pc-linux-gnu/6.5.0/../../../../x86_64-pc-linux-gnu/bin/ld: /var/tmp/portage/app-editors/laidout-9999/work/laidout-9999/laxkit/lax/interfaces/textonpathinterface.cc:992: undefined reference to `Laxkit::RefPtrStack<LaxInterfaces::PathsData>::RefPtrStack(char)'
/usr/lib/gcc/x86_64-pc-linux-gnu/6.5.0/../../../../x86_64-pc-linux-gnu/bin/ld: /var/tmp/portage/app-editors/laidout-9999/work/laidout-9999/laxkit/lax/interfaces/textonpathinterface.cc:1183: undefined reference to `Laxkit::RefPtrStack<LaxInterfaces::SomeData>::RefPtrStack(char)'
```

Tried with gcc 10.2, 7.5.0 and 6.5.0.

I believe this has something to do with changed default behaviour regarding template class instantiation in c++? Couldn't find the right syntax, but I'd wager the fix would be something like making sure that template class RefPtrStack<LaxInterfaces::SomeData>::RefPtrStack(char) et al gets instantiated somewhere.

Gcc 5 isn't available on Gentoo anymore, so I haven't tested with that.

**Comment by tomlechner** [comment]
Yes. I just updated Laxkit so hopefully this won't happen anymore.

It seems like each gcc release makes it more difficult to be casual about when my templates get built and linked! I've usually done templates with two files, a source and a header, which makes the resulting executable slightly smaller, and compiles slightly faster, but as time goes on, it keeps giving me headaches like this.. I guess it's time to just use header only templates!

**Comment by madsl** [comment]
Nice! Sorry for using this same bugreport for this, but there's also this in laidout git:

```
objecttree.cc: In member function ‘virtual int Laidout::ObjectTreeWindow::Event(const Laxkit::EventData*, const char*)’:
objecttree.cc:319:7: error: ‘viewport’ was not declared in this scope; did you mean ‘InViewport’?
  319 |   if (viewport && domain == InViewport) {
      |       ^~~~~~~~
      |       InViewport
objecttree.cc:312:24: warning: unused variable ‘sm’ [-Wunused-variable]
  312 |   const SimpleMessage *sm = dynamic_cast<const SimpleMessage*>(data);
      |                        ^~
```

I just commented it out, using this patch:
```
--- a/src/ui/objecttree.cc      2021-03-23 20:04:26.503716653 +0100
+++ b/src/ui/objecttree.cc      2021-03-23 20:05:32.485071828 +0100
@@ -316,9 +316,9 @@
                //DBG if (!viewport) return 0;
                //DBG viewport->PostMessage(sm->str);
 
-               if (viewport && domain == InViewport) {
+               // if (viewport && domain == InViewport) {
                        // viewport->Select(sm->str);
-               }
+               // }
 
                return 0;

```

**Comment by madsl** [comment]
Managed to create an ebuild now :)

https://gitlab.com/madsl/fol4/-/tree/master/app-editors/laidout

**Comment by tomlechner** [comment]
Arrr.. Must have forgotten to commit something! Usually the CI emails me when the repo can't compile, I must have missed it. Just committed a fix, so you shouldn't have to patch it. That ObjectTree thing is very much a work in progress!

That's great it builds! I must confess I don't know anything about Gentoo or ebuilds. If you think your ebuild setup is stable enough, I'd be happy to put a little blurb about it on laidout.org.

**Comment by madsl** [comment]
Cool! I've fixed the ebuild now so it doesn't have the viewport patch, and also added a version number override (-9999 versions in Gentoo means that the package is a live package built from newest sources). If these fixes makes it into 0.097.1 it's easy to create an ebuild for that too.

Instructions for installing in Gentoo:
Add the fol4 repository either by eselect: `eselect repository enable fol4` or by layman: `layman -fa fol4`
Install it by running `emerge -av laidout`

**Comment by madsl** [comment]
Fixed by 1f8d9efd1fd8c471f50a38d7f20a0b6a31d83a35 and https://github.com/Laidout/laxkit/commit/feb2aedbc2b15c6f29c0df876f898f19638ed5b4
